File: PQC/DotNetCore/MyMsPQC/TestMyMsPQC/KeyPairInfo.cs



namespace michele.natale.TestMsPqcs;


using Pointers;


public class KeyPairInfo : IDisposable
{
  public byte[] PublicKey { get; private set; } = [];
  public bool IsDisposed { get; private set; } = false;
  public UsIPtr<byte> PrivateKey { get; private set; } = UsIPtr<byte>.Empty;


  public KeyPairInfo(ReadOnlySpan<byte> pubkey, ReadOnlySpan<byte> privkey)
  {
    this.PublicKey = pubkey.ToArray();
    this.PrivateKey = new UsIPtr<byte>(privkey);
  }

  public KeyPairInfo(ReadOnlySpan<byte> pubkey, UsIPtr<byte> privkey)
  {
    this.PrivateKey = privkey.Copy;
    this.PublicKey = pubkey.ToArray();
  }

  public void Clear()
  {
    this.PrivateKey.Dispose();
    Array.Clear(this.PublicKey);

    this.PublicKey = [];
    this.PrivateKey = UsIPtr<byte>.Empty;
  }

  protected virtual void Dispose(bool disposing)
  {
    if (!this.IsDisposed)
    {
      if (disposing)
        this.Clear();
      this.IsDisposed = true;
    }
  }

  ~KeyPairInfo() =>
   Dispose(disposing: false);


  public void Dispose()
  {
    Dispose(disposing: true);
    GC.SuppressFinalize(this);
  }
}

File: PQC/DotNetCore/MyMsPQC/TestMyMsPQCVb/KeyPairInfo.vb
Option Strict On
Option Explicit On


Imports michele.natale.Pointers

Namespace michele.natale.TestMsPqcs
  Public Class KeyPairInfo
    Implements IDisposable

    Public Property PublicKey As Byte() = Nothing
    Public Property IsDisposed As Boolean = False
    Public Property PrivateKey As UsIPtr(Of Byte) = UsIPtr(Of Byte).Empty

    Public Sub New(pubkey As Byte(), privkey As Byte())
      Me.PublicKey = pubkey.ToArray()
      Me.PrivateKey = New UsIPtr(Of Byte)(privkey)
    End Sub

    Public Sub New(pubkey As Byte(), privkey As UsIPtr(Of Byte))
      Me.PrivateKey = privkey.Copy
      Me.PublicKey = pubkey.ToArray()
    End Sub

    Public Sub Clear()
      Me.PrivateKey.Dispose()
      Array.Clear(Me.PublicKey)

      Me.PublicKey = Nothing
      Me.PrivateKey = UsIPtr(Of Byte).Empty
    End Sub

    Protected Overridable Sub Dispose(disposing As Boolean)
      If Not Me.IsDisposed Then
        If disposing Then Me.Clear()
        Me.IsDisposed = True
      End If
    End Sub

    Protected Overrides Sub Finalize()
      Me.Dispose(disposing:=False)
      MyBase.Finalize()
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
      Me.Dispose(disposing:=True)
      GC.SuppressFinalize(Me)
    End Sub
  End Class
End Namespace
